### PR TITLE
Update the Fixer.Io api endpoint and method of submitting the api key.

### LINKED ISSUE #36010: Fixer IO API Endpoint changed
### Preconditions and environment

- Magento version: 2.4.5
- Anything else that would help a developer reproduce the bug

### Steps to reproduce

> Simply if you now go to configure currency exchange using Fixer IO they will direct you to a new platform with a new API Endpoint https://apilayer.com/marketplace/fixer-api
> Admin - Stores - Currency Rates - Select - Fixex.io as Import service and click on Import

### Expected result

It imports currency rates successfully

### Actual result

The import is failing due to the new API keys only compatible with the new API Endpoint which must be updated/corrected from within M2

### Additional information

_No response_

### Release note

_No response_

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [X] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by richard-parnaby-king**
### Description (*)
Fixer.io has changed the api endpoint and method of submitting the api key. Utilising Fixer.io to update currency rates results in an invalid api key error message. This PR updates the api endpoint and method of submitting the api key.

### Manual testing scenarios (*)
Go to Stores > Currency Rates , select Fixer.io as service and click on "Import". An error will appear.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#36010

### Contribution checklist (*)
- [x]  Pull request has a meaningful description of its purpose
- [x]  All commits are accompanied by meaningful commit messages
- [x]  All new or changed code is covered with unit/integration tests (if applicable)
- [ ]  README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
- [ ]  All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by richard-parnaby-king** [comment]
@magento run all tests

**Comment by richard-parnaby-king** [comment]
@magento run Static Tests

**Comment by ihor-sviziev** [comment]
Hi @richard-parnaby-king,
Thank you for your contribution!
I'm closing this PR as a duplicate of https://github.com/magento/magento2/pull/35984 that was created earlier

**Comment by richard-parnaby-king** [comment]
@ihor-sviziev That pull request does not solve the issue. The fixer api requires the access key to be submitted as part of a custom header, not as part of the url query string, which is what this pull request resolves.

File: app/code/Magento/Directory/Model/Currency/Import/FixerIo.php [hunk 1]
```diff
@@ -23,8 +23,13 @@ class FixerIo extends AbstractImport
...
      */
-    public const CURRENCY_CONVERTER_URL = 'http://data.fixer.io/api/latest?access_key={{ACCESS_KEY}}'
-        . '&base={{CURRENCY_FROM}}&symbols={{CURRENCY_TO}}';
+    public const CURRENCY_CONVERTER_URL = 'https://api.apilayer.com/fixer/latest?symbols{{CURRENCY_TO}}' .
+                                          '&base={{CURRENCY_FROM}}';
+    
+    /**
+     * @var string
+     */
+    public const API_KEY_CONFIG_PATH = 'currency/fixerio/api_key';
```
Comment (ihor-sviziev) [inline app/code/Magento/Directory/Model/Currency/Import/FixerIo.php]: Let's make it private due to backward compatibility reasons
```suggestion
    private const API_KEY_CONFIG_PATH = 'currency/fixerio/api_key';
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
@richard-parnaby-king please review and fix failing tests

**Review by Den4ik — CHANGES_REQUESTED** [review]
I believe that use of access key should be also removed at lines 114-115

```
url = str_replace(
            ['{{ACCESS_KEY}}', '{{CURRENCY_FROM}}', '{{CURRENCY_TO}}'],
            [$accessKey, $currencyFrom, $currenciesStr],
            self::CURRENCY_CONVERTER_URL
);
```

**Comment by richard-parnaby-king** [comment]
@ihor-sviziev

Some of these tests do not revolve around the functionality of this pull request, for example there is a failing test for the Aws S3 remote filestorage component. 

Looking at the failing tests the only one concerning the FixerIo module is failing this test: testFetchRates (https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Directory/Test/Unit/Model/Currency/Import/FixerIoTest.php)

The error message is:
```
Expectation failed for method name is "getValue" when invoked zero or more times
Parameter 0 for invocation #1 Magento\Framework\App\Config\ScopeConfigInterface::getValue('currency/fixerio/api_key', 'store', null) does not match expected value.
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'currency/fixerio/timeout'
+'currency/fixerio/api_key'
```

Which is this bit of code:
```
        $this->scopeConfig->method('getValue')
            ->withConsecutive(
                ['currency/fixerio/api_key', 'store'],
                ['currency/fixerio/timeout', 'store']
            )
            ->willReturnOnConsecutiveCalls('api_key', 100);
```

Do I need to fix this or does it sit with the core team? If I need to fix it, what is the expected functionality?

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@richard-parnaby-king, please fix the related tests only. For instance, the one you said fails because of changes.

**Comment by richard-parnaby-king** [comment]
@magento run Unit Tests

**Comment by richard-parnaby-king** [comment]
https://github.com/magento run Unit Tests

**Comment by richard-parnaby-king** [comment]
https://github.com/magento run Unit Tests

**Comment by richard-parnaby-king** [comment]
@magento run Unit Tests

**Comment by richard-parnaby-king** [comment]
@magento run all tests

**Comment by richard-parnaby-king** [comment]
@magento run All Tests

**Comment by richard-parnaby-king** [comment]
@ihor-sviziev Failing tests are unrelated to this PR.

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests EE, WebAPI Tests

**Review by ihor-sviziev — APPROVED** [review]

**Review by Den4ik — APPROVED** [review]

**Comment by engcom-Lima** [comment]
@magento run Functional Tests EE

**Comment by engcom-Lima** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Lima** [comment]
Hi @richard-parnaby-king 

Thanks for your contribution and collaboration.

I have tried to reproduce the issue ,it is reproducible to me below is the screenshot for your reference:
Screenshot 2022-09-15 at 5 39 06 PM

But after applying PR changes I am getting error in UI as well as in logs as below:
**IN LOGS:**
```
[2022-09-15T11:51:32.612615+00:00] main.CRITICAL: Error: Call to undefined method Magento\Framework\HTTP\LaminasClient::getHeaders() in /opt/homebrew/var/www/magento2/app/code/Magento/Directory/Model/Currency/Import/FixerIo.php:176
Stack trace:
#0 /opt/homebrew/var/www/magento2/app/code/Magento/Directory/Model/Currency/Import/FixerIo.php(121): Magento\Directory\Model\Currency\Import\FixerIo->getServiceResponse('https://api.api...')
#1 /opt/homebrew/var/www/magento2/app/code/Magento/Directory/Model/Currency/Import/FixerIo.php(81): Magento\Directory\Model\Currency\Import\FixerIo->convertBatch(Array, 'INR', Array)
#2 /opt/homebrew/var/www/magento2/app/code/Magento/CurrencySymbol/Controller/Adminhtml/System/Currency/FetchRates.php(87): Magento\Directory\Model\Currency\Import\FixerIo->fetchRates()
#3 /opt/homebrew/var/www/magento2/lib/internal/Magento/Framework/Interception/Interceptor.php(58): Magento\CurrencySymbol\Controller\Adminhtml\System\Currency\FetchRates->execute()
```
**IN UI:**
Screenshot 2022-09-15 at 6 23 26 PM

I have tested it in local as well in git cloud instance as well but issue is same.

Since PR is not working fine. I am moving this PR to **Changes requested**.

Thanks

**Review by engcom-Lima — CHANGES_REQUESTED** [review]
Hi @richard-parnaby-king 

Thanks for your contribution and collaboration.

I have tried to reproduce the issue ,it is reproducible to me below is the screenshot for your reference:
Screenshot 2022-09-15 at 5 39 06 PM

But after applying PR changes I am getting error in UI as well as in logs as below:
**IN LOGS:**
```
[2022-09-15T11:51:32.612615+00:00] main.CRITICAL: Error: Call to undefined method Magento\Framework\HTTP\LaminasClient::getHeaders() in /opt/homebrew/var/www/magento2/app/code/Magento/Directory/Model/Currency/Import/FixerIo.php:176
Stack trace:
#0 /opt/homebrew/var/www/magento2/app/code/Magento/Directory/Model/Currency/Import/FixerIo.php(121): Magento\Directory\Model\Currency\Import\FixerIo->getServiceResponse('https://api.api...')
#1 /opt/homebrew/var/www/magento2/app/code/Magento/Directory/Model/Currency/Import/FixerIo.php(81): Magento\Directory\Model\Currency\Import\FixerIo->convertBatch(Array, 'INR', Array)
#2 /opt/homebrew/var/www/magento2/app/code/Magento/CurrencySymbol/Controller/Adminhtml/System/Currency/FetchRates.php(87): Magento\Directory\Model\Currency\Import\FixerIo->fetchRates()
#3 /opt/homebrew/var/www/magento2/lib/internal/Magento/Framework/Interception/Interceptor.php(58): Magento\CurrencySymbol\Controller\Adminhtml\System\Currency\FetchRates->execute()
```
**IN UI:**
Screenshot 2022-09-15 at 6 23 26 PM

I have tested it in local as well in git cloud instance as well but issue is same.

Since PR is not working fine. I am moving this PR to **Changes requested**.

Thanks

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: app/code/Magento/Directory/Test/Unit/Model/Currency/Import/FixerIoTest.php [hunk 2]
```diff
@@ -110,6 +111,8 @@ public function testFetchRates(): void
             ->willReturnSelf();
         $httpClient->method('setMethod')
             ->willReturnSelf();
+        $httpClient->method('getHeaders')
+            ->willReturnSelf();
```
Comment (ihor-sviziev) [inline app/code/Magento/Directory/Test/Unit/Model/Currency/Import/FixerIoTest.php:115]: Maybe we should check if method setHeaders was executed with correct params?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by richard-parnaby-king** [comment]
@magento give me test instance

**Comment by richard-parnaby-king** [comment]
@magento give me test instance

**Comment by richard-parnaby-king** [comment]
@magento give me test instance

**Comment by richard-parnaby-king** [comment]
@magento give me test instance

**Comment by richard-parnaby-king** [comment]
@magento give me test instance

**Comment by ishakhsuvarov** [comment]
Looks like requested changes were not implemented or discussed in a long time, hence I am closing this PR for now.

Please feel free to reopen if there is any additional work done on this.

**Pull request closed without merging**